File: src/purpose/midwife.py
"""产婆 protocol + structural (offline) implementation.

A Midwife does not produce 志 — it only midwifes. Given a concern
and a chosen lens, it offers Socratic questions and (optionally) candidate
simplifications. The user is always the one who chooses or writes the
next, simpler concern.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from purpose.core import Journey
from purpose.lenses import LENSES, LensType


@dataclass(frozen=True, slots=True)
class Proposal:
    """What a midwife offers for one turn."""

    socratic: str
    candidates: tuple[str, ...] = ()
    note: str = ""
    anchor_suggestion: str | None = None


class Midwife(Protocol):
    name: str

    def propose(self, journey: Journey, lens: LensType) -> Proposal: ...


class StructuralMidwife:
    """Pure offline midwife: returns only the Socratic prompt for the lens.

    The user is the one who formulates the simpler concern. This is the
    baseline against which the LLM midwife is compared.
    """

    name = "structural"

    def propose(self, journey: Journey, lens: LensType) -> Proposal:
        spec = LENSES[lens]
        note = (
            "结构化模式：这里只给出苏格拉底式追问。"
            "化简后的陈述由你写下——请尽量短、更接近本质。"
        )
        if spec.is_terminal:
            note = "结构化模式：如果你能凭直觉判断当前陈述的方向，写下它，立为志。"
        return Proposal(socratic=spec.socratic, candidates=(), note=note)

File: docs/midwife.js
// 结构化产婆：给出苏格拉底式追问 + 模式化的提示语。
// LLM 产婆（v0.3）会替换这个函数，但接口保持一致：
//   propose(reading, lens) → { socratic, note }

import { LENS_BY_ID } from "./lenses.js";

export function structuralPropose(reading, lensId) {
  const lens = LENS_BY_ID[lensId];
  if (!lens) {
    return {
      socratic: "（未知透镜。）",
      note: "",
    };
  }
  if (lens.terminal) {
    return {
      socratic: lens.socratic,
      note: "如果你能凭直觉判断当前陈述就是这本书要回答的问题，写下它，立为读志。",
    };
  }
  return {
    socratic: lens.socratic,
    note: "写下化简后的一句陈述——要比当前更短、更接近本质。",
  };
}
